### ISSUE TITLE
Incorrect Regex matching result in compiled mode

### ISSUE THREAD

**Issue by veanes** [dotnet/runtime#52612]
Reproduce in .NET5.0.
The code below will print the following where False is an error:
False
True

```c#
string input = "this string has .digest as a substring";
Regex reC = new(@"[^a-z0-9]etag|[^a-z0-9]digest", RegexOptions.Compiled);
Regex re = new(@"[^a-z0-9]etag|[^a-z0-9]digest");
Console.WriteLine(reC.IsMatch(input));
Console.WriteLine(re.IsMatch(input));
```

**Comment by danmoseley** [comment]
This was a regression between .NET 3.1 and .NET 5.0. Unsurprisingly it's the same in .NET 6.0 previews.

**Comment by stephentoub** [comment]
> per @stephentoub "Seems likely this is a problem in either the analysis code that looks for what can be searched for at the beginning of a match

I just took a very quick look, and the expression asserts when run with a debug binary. The problem is indeed the analysis code, trying to merge two sets that the code doesn't support merging. The simple fix is to just bail on the analysis if the sets can't be merged (rather than just asserting). More complicated fix is to improve what sets can be merged.

I can take care of it when I'm back in a couple weeks if no one gets to it first.

**Comment by danmoseley** [comment]
With respect to servicing, we've apparently gotten no reports since we shipped this last Nov in 5.0. On the other hand, issues like this can be insidious, a little like bad codegen - when you upgrade to 5.0, in some cases your patterns may not match some inputs that can be easy to overlook. My suggestion is to let this fix bake for a month or so then consider servicing it.
